### PR TITLE
fix importAttributes is not avaliable in old node versions

### LINKED ISSUE #225: `importAttributes` not available in old node versions
reported by @weiatresvu here https://github.com/apollographql/vscode-graphql/pull/204#pullrequestreview-2361756806

> It was ⁠context.importAssertions. So, in some older versions of Node.js, context importAttributes will be undefined. Unfortunately, Cursor is using Node.js 20.9.0 in their latest version.
> 
> See, [nodejs.org/api/module.html#hooks:~:text=v21.0.0%2C v20.10.0%2C v18.19.0](https://nodejs.org/api/module.html#hooks:~:text=v21.0.0%2C%20v20.10.0%2C%20v18.19.0)

This would also impact VSCode before 1.92 (see [chart](https://github.com/ewanharris/vscode-versions), our current minimum is 1.90) so we will need a workaround for that.

@weiatresvu would you be open for a PR on this? :)

### PR DESCRIPTION
**Pull request by yesmeck**
fix #225 

Tested with VSCode 1.91.1 (Node.js 20.9.0) and VSCode 1.94.0 (Node.js 20.16.0).

### CODE REVIEW

**Comment by phryneas** [comment]
I'm merging in `main`, as I've added a CI run for 1.90.0 there.

File: src/language-server/config/cache-busting-resolver.js [hunk 1]
```diff
@@ -41,13 +80,19 @@ async function resolve(specifier, context, nextResolve) {
...
-  if (context.importAttributes.as !== "cachebust") {
+  const importAttributes = resolveImportAttributes(context);
+  const [as, format] = importAttributes.as.split(":");
+  if (as !== "cachebust") {
     return nextLoad(url, context);
   }
+  const contents =
+    "contents" in importAttributes
+      ? importAttributes.contents
+      : Object.keys(importAttributes)[1];
```
Comment (yesmeck) [inline src/language-server/config/cache-busting-resolver.js]: It’s pretty weird here is that the second key of importAssertions is the content of the module.

Comment (phryneas) [reply src/language-server/config/cache-busting-resolver.js]: Yes, seriously. But if it works it is what it is - this only applies to a small version range and will go away at some point in the future 🤷

File: src/language-server/config/cache-busting-resolver.js [hunk 2]
```diff
@@ -3,6 +3,38 @@ const { pathToFileURL } = require("node:url");
 
 /** @import { ResolveContext, ResolutionResult, LoadResult, ImportContext } from "./cache-busting-resolver.types" */
```
Comment (phryneas) [inline src/language-server/config/cache-busting-resolver.js:5]:
```suggestion
/** @import { ResolveContext, ResolutionResult, LoadResult, ImportContext,ImportAttributes } from "./cache-busting-resolver.types" */
```

File: src/language-server/config/cache-busting-resolver.js [hunk 3]
```diff
@@ -1,7 +1,38 @@
 // @ts-check
 const { pathToFileURL } = require("node:url");
 
-/** @import { ResolveContext, ResolutionResult, LoadResult, ImportContext } from "./cache-busting-resolver.types" */
+/** @import { ResolveContext, ResolutionResult, LoadResult, ImportContext, ImportAttributes } from "./cache-busting-resolver.types" */
+/**
+ * importAssertions was renamed to importAttributes in newer versions of Node.js.
+ *
+ * @param {ResolveContext|ImportContext} context
+ * @returns {boolean}
```
Comment (phryneas) [inline src/language-server/config/cache-busting-resolver.js]:
```suggestion
 * @returns {context is {importAttributes: ImportAttributes}}
```

File: src/language-server/config/cache-busting-resolver.js [hunk 4]
```diff
@@ -21,14 +52,16 @@ function bustFileName(specifier) {
  * @returns {Promise<ResolutionResult>}
  */
 async function resolve(specifier, context, nextResolve) {
-  if (context.importAttributes.as !== "cachebust") {
+  const importAttributes = resolveImportAttributes(context);
+  const [as, format] = importAttributes.as.split(":");
```
Comment (phryneas) [inline src/language-server/config/cache-busting-resolver.js]: This might need to move out into another function to make TS happy here

**Comment by phryneas** [comment]
Generally, could you please add some comment to document that once we hit a minimum of 1.92, most of these workarounds can be removed again, and reference the issue you started?
So we don't confuse future maintainers :)

File: src/language-server/config/cache-busting-resolver.js [hunk 5]
```diff
@@ -21,14 +52,16 @@ function bustFileName(specifier) {
...
+  if (as !== "cachebust") {
     return nextResolve(specifier, context);
   }
```
Comment (phryneas) [inline src/language-server/config/cache-busting-resolver.js]:
```suggestion
  const importAttributes = resolveImportAttributes(context);
  if (!importAttributes.as) {
    return nextResolve(specifier, context);
  }
  const [as, format] = importAttributes.as.split(":");
  if (as !== "cachebust" || (format !== "module" && format !== "commonjs")) {
    return nextResolve(specifier, context);
  }
```

File: src/language-server/config/cache-busting-resolver.js [hunk 6]
```diff
@@ -41,13 +74,19 @@ async function resolve(specifier, context, nextResolve) {
  * @returns {Promise<LoadResult>}
  */
 async function load(url, context, nextLoad) {
-  if (context.importAttributes.as !== "cachebust") {
+  const importAttributes = resolveImportAttributes(context);
+  const [as, format] = importAttributes.as.split(":");
+  if (as !== "cachebust") {
     return nextLoad(url, context);
```
Comment (phryneas) [inline src/language-server/config/cache-busting-resolver.js]:
```suggestion
  const importAttributes = resolveImportAttributes(context);
  if (!importAttributes.as) {
    return nextLoad(url, context);
  }
  const [as, format] = importAttributes.as.split(":");
  if (as !== "cachebust" || (format !== "module" && format !== "commonjs")) {
    return nextLoad(url, context);
```

Comment (phryneas) [reply src/language-server/config/cache-busting-resolver.js]: I think these two changes should make TS happy here

File: src/language-server/config/cache-busting-resolver.js [hunk 7]
```diff
@@ -53,8 +77,8 @@ function bustFileName(specifier) {
  */
 async function resolve(specifier, context, nextResolve) {
   const importAttributes = resolveImportAttributes(context);
-  const [as, format] = importAttributes.as.split(":");
-  if (as !== "cachebust") {
+  const format = resolveConfigFormat(importAttributes);
```
Comment (phryneas) [inline src/language-server/config/cache-busting-resolver.js:80]: I like this a lot better than my suggestion :)

File: src/language-server/config/cache-busting-resolver.js (hunk 5 above)
Comment (yesmeck) [reply src/language-server/config/cache-busting-resolver.js]: I've moved it out to a function.

**Comment by phryneas** [comment]
Okay, once the tests come back happy all that's left is a changeset - could you please run `npx changeset` and add a changeset for a patch-level change? :)

**Comment by phryneas** [comment]
Huh, that test runs fine for me locally. I'll investigate.

**Comment by yesmeck** [comment]
It seems we can't set `contents` on `importAssertions`. It's so weird that it was working for me.

**Comment by phryneas** [comment]
This is just super weird :(
Any more ideas?
Putting `contents` into `as` too? 😢

**Comment by yesmeck** [comment]
Can you rerun CI? It works for me locally now. We need to set both `⁠format` and `⁠contents` on the assert object, even though we can’t access the format from ⁠importAssertions, and the order of ⁠format and ⁠contents as keys also matters.

**Review by phryneas — APPROVED** [review]
And we're green.
Thank you *very* much for your hard work with this!

**Comment by phryneas** [comment]
Released as https://github.com/apollographql/vscode-graphql/releases/tag/v2.3.5

**Pull request merged**